src/core/exec-invoke.c

```diff
@@ -2563,12 +2617,41 @@ static int setup_exec_directory(
                 if (params->runtime_scope != RUNTIME_SCOPE_SYSTEM)
                         continue;
 
-                /* Then, change the ownership of the whole tree, if necessary. When dynamic users are used we
+                idmapping_supported = is_idmapping_supported(pp ?: p);
+                if (idmapping_supported < 0)
+                        goto fail;
+
+                log_debug("ID-mapping is%ssupported for exec directory %s", idmapping_supported ? " " : " not ", pp ?: p);
```

Comment:
thre's a lot of `pp ?: p` being used here now, please introduce a local variable (maybe placed before the chmod_and_chown() call above, that does this once. maybe: `const char *target_dir = pp ? p;`